### ISSUE TITLE
Consistent hash exchange should keep same distribution after server restart

### ISSUE THREAD

**Issue by acogoluegnes** [rabbitmq/rabbitmq-consistent-hash-exchange#32]
See https://groups.google.com/d/msg/rabbitmq-users/qHHptwZoxj8/XQbgGa8XBgAJ to reproduce.

**Comment by acogoluegnes** [comment]
The [`$2` number](https://github.com/rabbitmq/rabbitmq-consistent-hash-exchange/blob/master/src/rabbit_exchange_type_consistent_hash.erl#L77) that is matched against the hash to choose the bucket is [regenerated and "re-randomized" after each start](https://github.com/rabbitmq/rabbitmq-consistent-hash-exchange/blob/master/src/rabbit_exchange_type_consistent_hash.erl#L151). That's why buckets aren't consistent across restarts.

**Comment by michaelklishin** [comment]
This has been the case [from day one](https://github.com/rabbitmq/rabbitmq-consistent-hash-exchange/commit/23c33738c08d0a8baa0715fa85d5c344ad9931f7).

**Comment by acogoluegnes** [comment]
Weird nobody noticed before today.

**Comment by rai69** [comment]
I see that this issue is closed a long time ago, but with version 3.8.2 of RabbitMQ and the consistent-hash-exchange plugin I can reproduce the problem that after a restart the distribution is not always the same. Sometime it is, sometimes it isn't. Would you like me to file a new issue or provide information here?

**Comment by michaelklishin** [comment]
We would like to have a specific set of steps to reproduce. We won't investigate without one. The plugin has changed significantly since 2017, see #37 for example.

**Comment by michaelklishin** [comment]
One thing that can still affect the plugin even after #37 is binding changes between restarts or application deployments. Ultimately it's the applications that control the state of the consistent hashing ring via bindings, and changes to it can lead to changes in bucket distribution. Every binding has a "weight", for example.

Please put together an executable test with a set of steps that describes any administrative actions taken on the node (e.g. when it is restarted) and start a mailing list thread.

**Comment by rai69** [comment]
@michaelklishin I will be providing this today

**Comment by rai69** [comment]
My test is executed with a dotnet application that can be found here: https://github.com/rai69/rabbitmq-consisten-hash-test

**Setup RabbitMQ**
- An exchange of type topic named exchange-in
- An exchange of type x-consistent-hash named exchange-consumer connected to the exchange-in exchange
- 5 queues named consumer.0 through consumer.4 with a binding to exchange-consumer an

**Test**
To execute the test, following the following steps multiple times:

1. execute dotnet run .
2. examine the number of messages in the consumer queueus and make note of the number per queue
3. on the rabbitmq instance call: rabbitmqctl stop_app
4. on the rabbitmq instance call: rabbitmqctl stop_app
5. examine the number of messages in the consumer queueus these should be all 0
6. repeat from step 1 until the values in step 2 are different for 1 or more queues

**Results**
Initial run
![initial](https://github.com/rai69/rabbitmq-consisten-hash-test/raw/master/run-initial.png)
After a few test runs
![differen](https://github.com/rai69/rabbitmq-consisten-hash-test/raw/master/run-different-result.png)

**Problem**
After a restart it can't be garanteed that messsages will be sent to the same queue.

**Comment by rai69** [comment]
Can I get a response on this?

**Comment by michaelklishin** [comment]
@rai69 our team does not use GitHub issues for support and comments of a 3 year old issue is not where open source maintainers would prefer to troubleshoot things. We asked you to start a mailing list thread earlier.

There are two critically important aspects to using this plugin:

 * The hashing ring is controlled by the bindings of the exchange and their weights
 * Hashing is applied to the routing key, so routing keys and their distribution is another factor that affects routing

I don't see anything that would immediately stand out in your test.
This plugin has a reasonably [extensive test suite](https://github.com/rabbitmq/rabbitmq-consistent-hash-exchange/blob/master/test/rabbit_exchange_type_consistent_hash_SUITE.erl) which does not include E2E bindings. My best guess/hypothesis at the moment is that they can affect the state of the ring. Reproducing the examples from the test suite to compare would help prove this hypothesis right or wrong.

Another guess is that bindings are recovered in a different order, which means queues get different positions on the ring. The goal of this plugin is reasonably even distribution given a reasonably even distribution of routing keys (or other routing values), not guaranteed delivery to the very same queue. In fact it won't be the case in any workload with a changing topology.

**Comment by rai69** [comment]
@michaelklishin I missed the mailing thread part. Sorry for that. I will look into what you are saying and if necessary start a mailing thread.

**Comment by michaelklishin** [comment]
I will try to reproduce and see if the state of the ring changes between restarts. It is stored entirely in memory and reconstructed as bindings are recovered. Since there is no way to know what queue had what position in the ring before, this fully depends on binding recovery order and I have a guess bindings are loaded without any explicit ordering when they are recovered (but would need to confirm). Binding ordering in general is a fairly moot point as the [compound] primary key of a binding table is more or less the entire row.

So if we assume that binding recovery is not guaranteed to preserve order between boots, we cannot assume that the queue position on the hashing ring would be the same and thus we can guarantee reasonably even distribution but not routing locality.

**Comment by michaelklishin** [comment]
I have confirmed that my above hypothesis is correct. On one node boot, the state of the ring with the example from the `./examlpes` repository looks like this:

``` erlang
ets:tab2list(rabbit_exchange_type_consistent_hash_ring_state).
% => [{chx_hash_ring,{resource,<<"/">>,exchange,<<"chx">>},
% =>                 #{0 => {resource,<<"/">>,queue,<<"q1">>},
% =>                   1 => {resource,<<"/">>,queue,<<"q3">>},
% =>                   2 => {resource,<<"/">>,queue,<<"q3">>},
% =>                   3 => {resource,<<"/">>,queue,<<"q2">>},
% =>                   4 => {resource,<<"/">>,queue,<<"q4">>},
% =>                   5 => {resource,<<"/">>,queue,<<"q4">>}},
% =>                 6}]
```

after a restart, it looks like this:

``` erl
ets:tab2list(rabbit_exchange_type_consistent_hash_ring_state).
% =>[{chx_hash_ring,{resource,<<"/">>,exchange,<<"chx">>},
% =>                #{0 => {resource,<<"/">>,queue,<<"q1">>},
% =>                  1 => {resource,<<"/">>,queue,<<"q2">>},
% =>                  2 => {resource,<<"/">>,queue,<<"q4">>},
% =>                  3 => {resource,<<"/">>,queue,<<"q4">>},
% =>                  4 => {resource,<<"/">>,queue,<<"q3">>},
% =>                  5 => {resource,<<"/">>,queue,<<"q3">>}},
% =>                6}]
```

So all queues are there but their relative position in the ring is different, and therefore a routing key that hashes to the 4th bucket every time can still end up in a different queue after node restart.

**Comment by michaelklishin** [comment]
As much as we do not like discussions in comments of older issues, @rai69's question has led us to discover #45, and a solution I have as a prototype even allows us to recover exchanges and bindings in a specific order. While this does not mean that the ring can be originally constructed in the same order — this is fully controlled by the applications — it hopefully will reduce the confusion for some users and cases.

We have also updated the docs. Thank you, @rai69, your example was a meaningful contribution to this plugin.

**Comment by rai69** [comment]
@michaelklishin thank you for the clarification and in the future I will not comment on older closed issues!
